trieste/acquisition/function/entropy.py

```diff
@@ -617,3 +621,231 @@ def __call__(self, x: TensorType) -> TensorType:
+        :param dataset: The data from the observer.
+        """
+        query_points = self._search_space.sample(num_samples=self._grid_size)
+        tf.debugging.assert_same_float_dtype([dataset.query_points, query_points])
+        query_points = tf.concat([dataset.query_points, query_points], 0)
+        query_points_on_top_fidelity = tf.concat(
+            [query_points[:, :-1], tf.ones_like(query_points[:, -1:])], -1
+        )
+        return self._min_value_sampler.sample(
+            model, self._num_samples, query_points_on_top_fidelity
+        )
+
+
+class mumbo(AcquisitionFunctionClass):
+    def __init__(self, model: MUMBOModelType, samples: TensorType):
+        r"""
+        The MUMBO acquisition function of :cite:`moss2021mumbo`, modified for objective
+        minimisation. This function calculates the information gain (or change in entropy) in the
+        distribution over the objective minimum :math:`y^*`, if we were to evaluate the objective
+        at a given point on a given fidelity level.
+
+        To speed up calculations, we use a trick from :cite:`Moss:2021` and use moment-matching to
+        calculate MUMBO's entropy terms rather than numerical integration.
+
+        :param model: The model of the objective function.
+        :param samples: Samples from the distribution over :math:`y^*`.
+        :return: The MUMBO acquisition function modified for objective
+            minimisation. This function will raise :exc:`ValueError` or
+            :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+        :raise ValueError or tf.errors.InvalidArgumentError: If ``samples`` has rank less than two,
+            or is empty.
+        """
+        tf.debugging.assert_rank(samples, 2)
+        tf.debugging.assert_positive(len(samples))
+
+        self._model = model
+        self._samples = tf.Variable(samples)
+
+    def update(self, samples: TensorType) -> None:
+        """Update the acquisition function with new samples."""
+        tf.debugging.assert_rank(samples, 2)
+        tf.debugging.assert_positive(len(samples))
+        self._samples.assign(samples)
+
+    @tf.function
+    def __call__(self, x: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(x, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+
+        top_fidelity_idx = self._model.num_fidelities - 1
+        x_on_top_fidelity = tf.concat(
+            [tf.squeeze(x, -2)[:, :-1], top_fidelity_idx * tf.ones_like(tf.squeeze(x, -2)[:, -1:])],
+            -1,
+        )
+
+        fmean, fvar = self._model.predict(x_on_top_fidelity)
+        fsd = tf.clip_by_value(
+            tf.math.sqrt(fvar), CLAMP_LB, fmean.dtype.max
+        )  # clip below to improve numerical stability
+        ymean, yvar = self._model.predict_y(tf.squeeze(x, -2))
+        cov = self._model.covariance_with_top_fidelity(tf.squeeze(x, -2))
```

Comment:
Is it worth just doing this squeeze once and storing the output, rather than 3 times?